node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -207,14 +235,94 @@ impl OrderingProvider {
 	}
 }
 
+async fn send_message_fatal<Sender, Response>(
+	sender: &mut Sender,
+	message: impl Into<AllMessages>,
+	receiver: oneshot::Receiver<std::result::Result<Response, ChainApiError>>,
+) -> FatalResult<Response>
+where
+	Sender: SubsystemSender,
+{
+	sender.send_message(message.into()).await;
+
+	receiver
+		.await
+		.map_err(|_| Fatal::ChainApiSenderDropped)?
+		.map_err(Fatal::ChainApi)
+}
+
 async fn get_block_number(
 	sender: &mut impl SubsystemSender,
 	relay_parent: Hash,
 ) -> FatalResult<Option<BlockNumber>> {
 	let (tx, rx) = oneshot::channel();
-	sender.send_message(ChainApiMessage::BlockNumber(relay_parent, tx).into()).await;
+	send_message_fatal(sender, ChainApiMessage::BlockNumber(relay_parent, tx), rx).await
+}
+
+async fn get_finalized_block_hash(sender: &mut impl SubsystemSender) -> FatalResult<Option<Hash>> {
+	let (number_tx, number_rx) = oneshot::channel();
+	let number =
+		send_message_fatal(sender, ChainApiMessage::FinalizedBlockNumber(number_tx), number_rx)
+			.await?;
+
+	let (hash_tx, hash_rx) = oneshot::channel();
+	send_message_fatal(sender, ChainApiMessage::FinalizedBlockHash(number, hash_tx), hash_rx).await
+}
+
+/// Returns ancestor hashes of `head` in the descending order, stopping
+/// either at `last_observed_hash` or the genesis block.
+///
+/// Suited specifically for querying non-finalized chains, thus doesn't
+/// rely on block numbers.
+///
+/// The caller must ensure that `last_observed_hash` is indeed an ancestor
```

Comment:
This is in no way guaranteed right now. We should always at least stop at the first finalized block as a safeguard - even when using an LRU and always putting the latest finalized block in there, this would be too risky.